httpx/content.py

```diff
@@ -0,0 +1,162 @@
+import typing
+from json import dumps as json_dumps
+from urllib.parse import urlencode
+
+from .multipart import multipart_encode
+
+RequestData = typing.Union[dict, str, bytes, typing.AsyncIterator[bytes]]
+
+RequestFiles = typing.Dict[
+    str,
+    typing.Union[
+        # file (or str)
+        typing.Union[typing.IO[typing.AnyStr], typing.AnyStr],
+        # (filename, file (or str))
+        typing.Tuple[
+            typing.Optional[str], typing.Union[typing.IO[typing.AnyStr], typing.AnyStr],
+        ],
+        # (filename, file (or str), content_type)
+        typing.Tuple[
+            typing.Optional[str],
+            typing.Union[typing.IO[typing.AnyStr], typing.AnyStr],
+            typing.Optional[str],
+        ],
+    ],
+]
+
+
+class RequestContent:
+    """
+    Base class for request content.
+    Defaults to a "no request body" implementation.
+    """
+
+    def get_headers(self) -> typing.Dict[str, str]:
+        return {}
+
+    def can_rewind(self) -> bool:
+        return True
+
+    async def __aiter__(self) -> typing.AsyncIterator[bytes]:
+        yield b""
+
+    async def aread(self) -> bytes:
```

Comment:
Also, do we really want `.aread()` as a name here?

Thinking about `unasync`, this means that the sync version of this would be `def aread(self)` by default (unless `unasync` also supports removing any "a" prefix, but I don't think it does/should).

So maybe this should be `async def read(self) -> bytes` which would map to `def read(self) -> bytes` on the `SyncResponseContent` variant?

Or is that not the `unasync` use case you're thinking of here?

Reply:
Thinking about it we may not need `unasync` at all to support sync here, right? Kind of the point of this whole thing I suppose. 😄 We'll just implement `def read()` and `__iter__()`, right?

Reply:
> Thinking about it we may not need unasync at all to support sync here, right? Kind of the point of this whole thing I suppose. 😄 We'll just implement def read() and __iter__(), right?

Exactly, yes. `Request` and `Response` *won't* have two different class variants.